### PR TITLE
ARROW-13507: [R] LTO job on CRAN fails

### PR DESCRIPTION
**Pull request by jonkeane**
A test to see if we can (for now) build r-debug before using it

### CODE REVIEW

File: dev/tasks/docker-tests/azure.linux.yml [hunk 1]
```diff
@@ -45,6 +45,15 @@ jobs:
       git -C arrow submodule update --init --recursive
     displayName: Clone arrow
 
+  # only needed for r-lto job to use until https://github.com/wch/r-debug/pull/22 is merged
+  - script: |
+      git clone https://github.com/nealrichardson/r-debug.git
+      git checkout lto
+      cd r-debug
+      ./buildall.sh
```
Comment (nealrichardson) [inline dev/tasks/docker-tests/azure.linux.yml:53]: A possibly faster way to do this is to add a new job like for the r-sanitizer, and add dockerfile like https://github.com/nealrichardson/r-debug/blob/lto/r-debug-5/Dockerfile:

```
FROM wch1/r-debug

# RDlto: R built with --enable-lto
RUN wget https://raw.githubusercontent.com/nealrichardson/r-debug/lto/r-devel/buildR.sh
RUN buildR.sh lto
RUN RDlto -q -e 'install.packages("pak", repos = "https://r-lib.github.io/p/pak/dev")'
RUN RDlto -q -e 'pak::pkg_install(c("devtools", "Rcpp", "roxygen2", "testthat", "memoise", "rmarkdown"))'
```

Comment (nealrichardson) [reply dev/tasks/docker-tests/azure.linux.yml:53]: Actually, turns out R-hub has a docker image already: https://github.com/r-hub/rhub-linux-builders/blob/master/debian-gcc-devel-lto/Dockerfile

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-r_hub-debian-gcc-devel-lto-latest

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

This should fail with the CRAN error at https://www.stats.ox.ac.uk/pub/bdr/LTO/arrow.out

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by jonkeane** [comment]
I also ran this locally removing all of the `-fno-lto` flags that were added in ARROW-13199, and I get a segfault there as well (this is at the same point as the crossbow build above which tries to remove that flag only when we know we're on LTO):

```
16: tools:::.test_load_package("arrow", "/arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new")
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault
ERROR: loading failed
```

So it looks like we need to do something more to be able to build/link on the CRAN LTO check.

File: r/inst/build_arrow_static.sh [hunk 2]
```diff
@@ -75,6 +79,7 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
     -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON \
     -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON \
     -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD:-ON} \
+    -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=${ARROW_LTO:-ON} \
```
Comment (cpcloud) [inline r/inst/build_arrow_static.sh:82]: Should this be `${ARROW_LTO:-$ARROW_DEFAULT_PARAM}`? As it currently is, it looks like LTO will always be turned on.

File: r/inst/build_arrow_static.sh [hunk 3]
```diff
@@ -45,6 +45,10 @@ else
   ARROW_DEFAULT_PARAM="OFF"
 fi
 
+if echo "$ARROW_R_CXXFLAGS" | grep -q "flto"; then
+  ARROW_LTO="ON"
```
Comment (cpcloud) [inline r/inst/build_arrow_static.sh:49]: Should this have an `else` clause setting `ARROW_LTO` to `"OFF"`?

File: r/inst/build_arrow_static.sh (hunk 2 above)
Comment (jonkeane) [reply r/inst/build_arrow_static.sh:82]: I've changed this to `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=${ARROW_LTO:-OFF}` Confusingly, `$ARROW_DEFAULT_PARAM` here is about adding extra features like compression libraries (based on `$LIBARROW_MINIMAL` above).

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**